velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -320,104 +423,91 @@ void HiveDataSink::close() {
   }
 }
 
-void HiveDataSink::ensureSingleWriter() {
-  if (writers_.empty()) {
-    appendWriter(std::nullopt);
+uint32_t HiveDataSink::ensureWriter(const HiveWriterId& id) {
+  auto it = writerIndexMap_.find(id);
+  if (it != writerIndexMap_.end()) {
+    return it->second;
   }
-}
-
-void HiveDataSink::ensurePartitionWriters() {
-  const auto numPartitions = partitionIdGenerator_->numPartitions();
-  const auto numWriters = writers_.size();
-
-  VELOX_CHECK_LE(numWriters, numPartitions);
-
-  if (numWriters < numPartitions) {
-    writers_.reserve(numPartitions);
-    writerInfo_.reserve(numPartitions);
-    for (auto id = numWriters; id < numPartitions; ++id) {
-      appendWriter(partitionIdGenerator_->partitionName(id));
-    }
+  return appendWriter(id);
+}
+
+uint32_t HiveDataSink::appendWriter(const HiveWriterId& id) {
+  // Check max open writers.
+  VELOX_USER_CHECK_LE(
+      writers_.size(),
+      maxOpenWriters_,
+      "Exceeded limit of {} open writers",
```

Comment:
No need to include maxOpenWriters_ in the error message. It is already included.